### PR TITLE
Record probe data with all pusher Pusher

### PR DESCRIPTION
**Pull request by psychocoderHPC**
Add functionality to all pushers to record particle attribute probeB and probeE data.
PMacc: add meta function `executeIf` and `executeIfElse`.

All pushers record field data if a particle as the attribute `probeB` or `probeE`. 
There is no need anymore to add attributes, depending on which attributes are available to corresponding data will be recorded.

# Todo

 - [x] add doxygen to `executeIf` and `executeIfElse`
 - [x] update PIConGPU probe documentation
 - [x] rebase after #3715 is merged

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -41,7 +41,7 @@ pull-request-validation:
     - wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | apt-key add -
     - echo "deb http://apt.llvm.org/focal/ llvm-toolchain-focal-11 main" | tee -a /etc/apt/sources.list
     - apt update
-    - apt install -y clang-format-11
+    - DEBIAN_FRONTEND=noninteractive apt install -y clang-format-11
```
Comment (PrometheusPi) [inline .gitlab-ci.yml:44]: should not be part of this pr

Comment (psychocoderHPC) [reply .gitlab-ci.yml:44]: I cherry-picked this to pass the CI, this PR will be automatically removed after the final rebase against #3715

**Comment by wmles** [comment]
I can compile your branch just fine, but not running.
I even tried to check out your branch and clone an LCT example and get:
```
Unhandled exception of type 'St13runtime_error' with message '/p/home/jusers/goethel1/juwels/simulations/210803_TestImportOpmd/picongpu/thirdParty/cupla/alpaka/include/alpaka/event/EventUniformCudaHipRt.hpp(173) 'ret = ALPAKA_API_PREFIX(EventQuery)(event.m_spEventImpl->m_UniformCudaHipEvent)' returned error  : 'cudaErrorLaunchFailure': 'unspecified launch failure'!', terminating
terminate called after throwing an instance of 'std::runtime_error'
  what():  /p/home/jusers/goethel1/juwels/simulations/210803_TestImportOpmd/picongpu/thirdParty/cupla/alpaka/include/alpaka/event/EventUniformCudaHipRt.hpp(227) 'cudaEventSynchronize(event.m_spEventImpl->m_UniformCudaHipEvent)' returned error  : 'cudaErrorLaunchFailure': 'unspecified launch failure'!
srun: error: jwb0027: task 0: Aborted (core dumped)
```
this was on juwels, I will try on hemera presently. One could even look into the core dump if you like?

File: include/picongpu/particles/pusher/particlePusherFree.hpp [hunk 2]
```diff
@@ -52,6 +54,17 @@ namespace picongpu
                 using MomType = momentum::type;
                 MomType const mom = particle[momentum_];
 
+                const auto bField = functorBField(pos);
+                const auto eField = functorEField(pos);
```
Comment (psychocoderHPC) [inline include/picongpu/particles/pusher/particlePusherFree.hpp:58]: BUG: The guard size of the pusher must be adjusted if we access the fields to interpolate the fields!

**Comment by psychocoderHPC** [comment]
> I can compile your branch just fine, but not running.
> In the meantime I'm pretty sure it's not my fault - I even tried to check out your branch and clone an LCT example and get:

Do you change anything in the FoildCT example? I tried it myself and everything was running fine.
Never the less for the general PR I found an issue (forgot to handle the guard size for the stencil)

**Comment by wmles** [comment]
> > I can compile your branch just fine, but not running.
> > In the meantime I'm pretty sure it's not my fault - I even tried to check out your branch and clone an LCT example and get:
> 
> Do you change anything in the FoildCT example? I tried it myself and everything was running fine.
> Never the less for the general PR I found an issue (forgot to handle the guard size for the stencil)

I just see it runs on hemeras k20 queue. So this seems to be a juwels-specific issue, is this possible? I will try another time on juwels (yes, I hope that I did not change anything on the example the last time, but just to be even more sure)

**Comment by wmles** [comment]
> > > I can compile your branch just fine, but not running.
> > > In the meantime I'm pretty sure it's not my fault - I even tried to check out your branch and clone an LCT example and get:
> > 
> > 
> > Do you change anything in the FoildCT example? I tried it myself and everything was running fine.
> > Never the less for the general PR I found an issue (forgot to handle the guard size for the stencil)
> 
> I just see it runs on hemeras k20 queue. So this seems to be a juwels-specific issue, is this possible? I will try another time on juwels (yes, I hope that I did not change anything on the example the last time, but just to be even more sure)

I confirm - on juwels a paramSet compiled wich dev runs, and compiled with the new commit, breaks.
On hemera it runs on k20 and fwkt_v100 (compiled for k20)

**Comment by psychocoderHPC** [comment]
@wmles It is not running on juwles booster right? Do you tested foilCT from the pull request branch without any changes?
If you tested any other example there is the possibility that you run out of memory, therefore I am asking again.

The error message you posted can not be triggered by rhis PR.

**Comment by psychocoderHPC** [comment]
I will finish this PR in two weeks.

**Comment by sbastrakov** [comment]
> I will finish this PR in two weeks.

@wmles if this is needed more urgently I think I can take over.

File: include/picongpu/particles/pusher/particlePusherFree.hpp (hunk 2 above)
Comment (psychocoderHPC) [reply include/picongpu/particles/pusher/particlePusherFree.hpp:58]: False alert: The guard size depends on the interpolation function. If a probe has defined interpolation the guard is derived correctly.

```
    using ParticleFlagsProbes = MakeSeq_t<
        particlePusher<particles::pusher::Probe>,
        shape<UsedParticleShape>,
        interpolation<UsedField2Particle>>;
```

**Comment by psychocoderHPC** [comment]
ready for review:  @wmles issue can not be reproduced, the changes important for FoilLCT are also very small and should not produce the issue.

https://github.com/ComputationalRadiationPhysics/picongpu/pull/3714/files#diff-2187eb5b3207355fc3f0ec3ad97d38c8e214d58be25bb8c34c2fa198577ca381R108-R115

`cuda-memcheck` is also not reporting any invalid memory access.

File: include/pmacc/meta/ExecuteIf.hpp [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        /** Conditional functor execution
+         *
+         * Evaluate and execute a functor based on the conditional compile-time argument.
+         * The functor implementations must have valid C++ syntax but is only evaluated if executed.
+         *
+         * @tparam T_condition execute trueFunctor if the condition is true, else falseFunctor
+         * @tparam T_ConditionTrueFunctor functor type, the functor is only evaluated if T_condition is true
+         * @tparam T_ConditionFalseFunctor functor type, the functor is only evaluated if T_condition is false
+         * @tparam T_Args types of the arguments for the executed functor
+         * @param trueFunctor functor instance which called with the arguments args if T_condition is true
```
Comment (n01r) [inline include/pmacc/meta/ExecuteIf.hpp:99]: which *is* called
-> also in the line below

File: include/pmacc/meta/ExecuteIf.hpp [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        } // namespace detail
+
+        /** Conditional functor execution
+         *
+         * Evaluate and execute a functor based on the conditional compile-time argument.
+         * The functor implementations must have valid C++ syntax but is only evaluated if executed.
+         *
+         * @tparam T_condition execute trueFunctor if the condition is true, else no operation is performed
+         * @tparam T_ConditionTrueFunctor functor type, the functor is only evaluated if T_condition is true
+         * @tparam T_Args types of the arguments for trueFunctor call
```
Comment (n01r) [inline include/pmacc/meta/ExecuteIf.hpp:77]: just a technical question:

Is it unlikely that the `falseFunctor` will have possible arguments?
Or should there maybe be `T_ArgsOnTrue` and `T_ArgsOnFalse` for added generality?

**Review by n01r — CHANGES_REQUESTED** [review]
All the code looks good!
Just a technical question in the comments and the docs for the 'moving probes' needs to go to the tracer particles. :)

**Comment by n01r** [comment]
Ah, right, an @wmles will do the testing, he said.

**Comment by wmles** [comment]
> ready for review: @wmles issue can not be reproduced, the changes important for FoilLCT are also very small and should not produce the issue.

You tried to reproduce the issue on hemera? It does not occur on hemera, it should be reproduced on juwels.

And yes, I just tried it once more - new folder, new git clone to your topic-allPusherprobes-branch, correct which pic-create, creating from FoilLCT -> compiles fine, does not run. git checkout dev -> compiles again, runs.

File: include/pmacc/meta/ExecuteIf.hpp (hunk 4 above)
Comment (psychocoderHPC) [reply include/pmacc/meta/ExecuteIf.hpp:77]: The arguments should be the same for the false and true  functor.
You can use the lambda functor capture to pass different arguments into the functor.

**Comment by psychocoderHPC** [comment]
@wmles Thanks that you tested the branch again. I will test the PR next week on JWLES too and check what's going on.

File: docs/source/usage/workflows/tracerParticles.rst [hunk 5]
```diff
@@ -8,21 +8,78 @@ Tracer Particles
...
+      using ParticleFlagsProbeElectrons = MakeSeq_t<
+          particlePusher< particles::pusher::Boris >,
+          shape< UsedParticleShape >,
+          interpolation< UsedField2Particle >,
+          currentSolver::EmZ<UsedParticleCurrentSolver>
+      >;
+
+      using TracerElectron = Particles<
+          PMACC_CSTRING( "e_tracer" ),
+          ParticleFlagsProbeElectrons,
```
Comment (psychocoderHPC) [inline docs/source/usage/workflows/tracerParticles.rst:29]: ParticleFlagsProbeElectrons -> ParticleFlagsTracerElectrons

**Comment by psychocoderHPC** [comment]
The problem why it is crashing on JUWLES is that the compiler is generating strange code:

In case I remove the `probeE` attribute from the probe species in the foil example and remove all my changes from BorisPusher the compiler is generating the wrong code with a `trap`.
The trap is obviously stopping the kernel.
```
//picongpu/include/pmacc/../picongpu/particles/pusher/particlePusherProbe.hpp:111                     [&bField, &valueFunctor](auto&& par) { valueFunctor(par[probeB_], bField); },
	.loc 139 111 60
	mul.wide.s32 	%rd87, %r3, 12;
	add.s64 	%rd88, %rd16, %rd87;
	st.global.f32 	[%rd88+2048], %f93;
	st.global.f32 	[%rd88+2052], %f167;
	st.global.f32 	[%rd88+2056], %f188;
	trap;
```

**Comment by psychocoderHPC** [comment]
@n01r I added your suggestions.

This PR can now be reviewed and merged, the BUG on JUWELS booster is solved.

File: docs/source/usage/workflows/probeParticles.rst [hunk 6]
```diff
@@ -16,7 +16,7 @@ Self-consistently interacting particles are usually called :ref:`tracer particle
 Workflow
 """"""""
 
-* ``speciesDefinition.param``: create a species specifically for probes and add ``probeE`` and ``probeB`` attributes to it for storing interpolated fields
+* ``speciesDefinition.param``: create a non moving probe species, add ``probeE`` and ``probeB`` attributes to it for storing interpolated fields
```
Comment (n01r) [inline docs/source/usage/workflows/probeParticles.rst:19]: *non-moving* or *stationary*

File: docs/source/usage/workflows/probeParticles.rst [hunk 7]
```diff
@@ -114,3 +114,4 @@ Known Limitations
 .. warning::
 
    If the probe particles are dumped in the file output, the instantaneous fields they recorded will be one time step behind the last field update (since our runOneStep pushed the particles first and then calls the field solver).
+   Adding the attributes probeE or probeB to a species will increase the particle memory footprint by ``3 * sizeof(float_X)`` byte per attribute and particle.
```
Comment (n01r) [inline docs/source/usage/workflows/probeParticles.rst:117]: ``probeE`` or ``probeB`` (set text verbatim / code style)

File: docs/source/usage/workflows/tracerParticles.rst [hunk 8]
```diff
@@ -8,21 +8,78 @@ Tracer Particles
 Tracer particles are like :ref:`probe particles <usage-workflows-probeParticles>`, but interact self-consistenly with the simulation.
 They are usually used to visualize *representative* particle trajectories of a larger distribution.
 
+In PIConGPU each species can be a tracer particle species, this allows tracking fields along the particle trajectory.
```
Comment (n01r) [inline docs/source/usage/workflows/tracerParticles.rst:11]: *... which allows ...* or make it a new sentence

File: docs/source/usage/workflows/tracerParticles.rst [hunk 9]
```diff
@@ -8,21 +8,78 @@ Tracer Particles
...
+          ParticleFlagsTracerElectrons,
+          MakeSeq_t<
+              position< position_pic >,
+              momentum,
+              weighting,
+              particleId
+          >
+      >;
+
+* ``speciesDefinition.param``: create a species specifically for tracer particles which only stores ``probeE``, the electric interpolated field
```
Comment (n01r) [inline docs/source/usage/workflows/tracerParticles.rst:38]: *interpolated electric field*

File: docs/source/usage/workflows/tracerParticles.rst [hunk 10]
```diff
@@ -8,21 +8,78 @@ Tracer Particles
...
+      using ParticleFlagsTracerElectrons = MakeSeq_t<
+          particlePusher< particles::pusher::Boris >,
+          shape< UsedParticleShape >,
+          interpolation< UsedField2Particle >,
+          currentSolver::EmZ<UsedParticleCurrentSolver>
+      >;
 
-  * add the particle attribute ``particleId`` to your species' ``Particles< ... >`` class (third argument, ``T_Attributes``)
-  * optional: add ``fieldE`` and ``fieldB`` attributes to the species to store fields as in :ref:`probes <usage-workflows-probeParticles>` 
+      using TracerElectron = Particles<
```
Comment (n01r) [inline docs/source/usage/workflows/tracerParticles.rst:49]: Hmm, I wonder if it makes sense to have two code blocks for this.
Usually, when I have a tracer particle with E-field recording, I'd also like to have its ID to follow its path.
IMO one code block that adds both `particleId` and `probeE` is enough.

**Review by n01r — COMMENTED** [review]
I basically approve and just marked some nitpicky changes to documentation.
I will merge once they are addressed. :)

File: docs/source/usage/workflows/probeParticles.rst (hunk 7 above)
Comment (sbastrakov) [reply docs/source/usage/workflows/probeParticles.rst:117]: Maybe we should also clarify that the memory footprint is for this species. (Some new users may misunderstand as if it was for all species e.g. if all species had same attributes)

File: docs/source/usage/workflows/tracerParticles.rst (hunk 8 above)
Comment (sbastrakov) [reply docs/source/usage/workflows/tracerParticles.rst:11]: Should we call them `probes`, not `tracer` everywhere?

File: docs/source/usage/workflows/tracerParticles.rst (hunk 10 above)
Comment (sbastrakov) [reply docs/source/usage/workflows/tracerParticles.rst:49]: > Usually, when I have a tracer particle with E-field recording, I'd also like to have its ID to follow its path

Maybe it's worth putting this explicitly before the code, and then express in the code?

File: include/picongpu/param/speciesAttributes.param [hunk 11]
```diff
@@ -73,10 +73,16 @@ namespace picongpu
...
 
-    //! interpolated electric field with respect to particle shape
+    /** interpolated electric field with respect to particle shape
+     *
+     * Attribute can be added to any species.
+     */
     value_identifier(float3_X, probeE, float3_X::create(0.));
 
-    //! interpolated electric field with respect to particle shape
+    /** interpolated electric field with respect to particle shape
```
Comment (sbastrakov) [inline include/picongpu/param/speciesAttributes.param:82]: `magnetic`

File: docs/source/usage/workflows/tracerParticles.rst (hunk 8 above)
Comment (n01r) [reply docs/source/usage/workflows/tracerParticles.rst:11]: Well, I like the conceptual separation between stationary probes that just get placed somewhere in an arbitrary distribution to record what's going on and the representatives of a moving, physical species that you can trace.
As long as it's clear that they both originate from the same concept of recording and just differ in what pusher they have.

File: include/pmacc/meta/ExecuteIf.hpp [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+         * @param trueFunctor functor instance which is called with the arguments args if T_condition is true
+         * @param args arguments forwarded to the executed functor
+         */
+        template<bool T_condition, typename T_ConditionTrueFunctor, typename... T_Args>
+        constexpr void executeIf(T_ConditionTrueFunctor&& trueFunctor, T_Args&&... args)
```
Comment (sbastrakov) [inline include/pmacc/meta/ExecuteIf.hpp:82]: This works in CUDA and HIP as the function is `constexpr`. But probably it is more generaly to add `HDINLINE` here and for the `IfElse` function

Comment (sbastrakov) [reply include/pmacc/meta/ExecuteIf.hpp:82]: Unrelated: i think `invokeIf` would be a  more suitable name as here we are running some function which is commonly named _invoking_, see [std::invoke](https://en.cppreference.com/w/cpp/utility/functional/invoke)

File: include/pmacc/meta/ExecuteIf.hpp [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+
+            /** Specialization if condition is true
+             *
+             * Execute trueFunctor passed to operator()
+             */
+            template<>
+            struct IfElse<true>
+            {
+                template<typename T_ConditionTrueFunctor, typename T_ConditionFalseFunctor, typename... T_Args>
+                void operator()(T_ConditionTrueFunctor&& trueFunctor, T_ConditionFalseFunctor&&, T_Args&&... args)
```
Comment (sbastrakov) [inline include/pmacc/meta/ExecuteIf.hpp:47]: This should be `HDINLINE` and the other specialization as well.

Comment (sbastrakov) [reply include/pmacc/meta/ExecuteIf.hpp:47]: Also offline discussion. How does it work at all for CUDA? If interpreted as written, this function can't be called on device at all. So does the compiler inline more now? Perhaps related to https://github.com/alpaka-group/alpaka/issues/1408#issuecomment-914880228

File: docs/source/usage/workflows/tracerParticles.rst (hunk 10 above)
Comment (psychocoderHPC) [reply docs/source/usage/workflows/tracerParticles.rst:49]: @n01r I kept the *ID* and *probe* separate. The id is used to follow a particle where the probe attribute is what the particle sees during the walk. I do not want the user thinking an *ID* is required for tracer species.

File: include/picongpu/particles/pusher/particlePusherAxel.hpp [hunk 14]
```diff
@@ -82,8 +84,17 @@ namespace picongpu
...
+                const auto eField = functorEField(pos);
+
+                // update probe field if particle contains required attributes
+                pmacc::meta::executeIf<pmacc::traits::HasIdentifier<T_Particle, probeB>::type::value>(
+                    [&bField](auto&& par) { par[probeB_] = bField; },
+                    particle);
+
+                pmacc::meta::executeIf<pmacc::traits::HasIdentifier<T_Particle, probeE>::type::value>(
+                    [&eField](auto&& par) { par[probeE_] = eField; },
+                    particle);
```
Comment (sbastrakov) [inline include/picongpu/particles/pusher/particlePusherAxel.hpp:97]: Also discussed offline a few days ago. This piece is probably better as a functor/function called from all pushers.

Comment (psychocoderHPC) [reply include/picongpu/particles/pusher/particlePusherAxel.hpp]: This will be part of a follow-up PR to avoid breaking the already well-tested implementation.

File: include/pmacc/meta/ExecuteIf.hpp (hunk 13 above)
Comment (psychocoderHPC) [reply include/pmacc/meta/ExecuteIf.hpp]: done

**Review by sbastrakov — APPROVED** [review]

**Comment by sbastrakov** [comment]
@n01r please re-review and accept changes if you are fine with the current version.

**Review by n01r — APPROVED** [review]
